front/loadObjects.js

```diff
@@ -1,44 +1,118 @@
+                
+        });
 
-                if(!entry.visible){
-                    scene.add(entry.mesh);
-                    entry.visible = true;
-                }
+        console.log('nb Visible', meshFromId.size, 'requestedIds', objectRequestedIds.size, 'nbToLoad', nbToLoad);
+    };
+    
+    function Zone(scene, camera, domElement) {
+        // query the rtree to know what building are needed
+        var nbToLoad = 0;
+
+        var position = camera.position;
+
+        var L = 2 * camera.position.z * Math.tan(Math.PI*camera.fov/(2*180));
+        var l = L * domElement.clientWidth / domElement.clientHeight;
+
+        var south = position.y - L/2;
+        var north = position.y + L/2;
+        var west = position.x - l/2;
+        var east = position.x + l/2;
+        
+        // ask for a little extra
+        west -= 100;
+        south -= 100;
+        east += 100;
+        north += 100;
+
+        var resultsBase = rTree.search([west, south, east, north]);
+        var resultsExtended = rTree.search([west-100, south-100, east+100, north+100]);
+
+        // in base and extended zones
+        var objectRequestedIdsBase = new Set(resultsBase.map(function(r){ return r[4].id; }));
+        var objectRequestedIdsExtended = new Set(resultsExtended.map(function(r){ return r[4].id; }));
+
+        // in extended zone ...
+        objectRequestedIdsExtended.forEach(function(id){
+            
+            // ... but not in scene => ask for buildings
+            if (!meshFromId.has(id)){
+                nbToLoad ++;
+                
+                getCityObject(id)
+                .then(function(object){
+                    // if also in base zone, add mesh to scene
+                    if (objectRequestedIdsBase.has(id)){
+                        var mesh = unpackBuilding(object);
+                        scene.add(mesh);
+                    }
+                });
             }
         });
 
-        // hide objects which are visible, but shouldn't
-        displayedObjectIds.forEach(function(id){
-            if(!objectToDisplayIds.has(id)){
-                var object = buildingMap.get(id);
-                object.visible = false;
-                scene.remove(object.mesh);
-            }    
+        var nbToHide = 0;
+
+        // in base zone but too far => hide
+        meshFromId.forEach(function(mesh, id){
+            if (infosFromMesh.get(mesh).type === 'building' && distance(mesh.position, camera.position) > 1000) {
+                nbToHide ++;
+                scene.remove(mesh);
+                infosFromMesh.delete(meshFromId.delete(id));
+            }
         });
 
+        console.log('extendedIds', objectRequestedIdsExtended.size, 'baseIds', objectRequestedIdsBase.size,  'buildingCache', buildingCache.size);
+        console.log('nbToLoad', nbToLoad, 'nbVisible', meshFromId.size, 'nbToHide', nbToHide);
     };
     
+    return {
+        fast: Fast,
+        zone: Zone
+    };
```

Comment:
These are use case-specific. What if a control needs a new type? Should we change the API?
I believe we shouldn't. We should provide one generic loading function and controls build on top of that.